sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,841 @@
+                if (syntaxNode is MethodDeclarationSyntax methodSyntax)
+                {
+                    if (methodSyntax.AttributeLists.Count > 0) // collect all methods with attributes - we will verify they are functions when we have access to symbols to get the full name
+                    {
+                        CandidateMethods.Add(methodSyntax); 
+                    }
+                }
+            }
+        }
+
+        private static void WriteGetFunctionsMetadataAsyncMethod(IndentedTextWriter indentedTextWriter, SyntaxReceiver receiver, Compilation compilation)
+        {
+            indentedTextWriter.WriteLine("public Task<ImmutableArray<IFunctionMetadata>> GetFunctionMetadataAsync(string directory)");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+
+            // create list of IFunctionMetadata and populate it
+            indentedTextWriter.WriteLine("var metadataList = new List<IFunctionMetadata>();");
+            AddFunctionMetadataInfo(indentedTextWriter, receiver, compilation);
+            indentedTextWriter.WriteLine("return Task.FromResult(metadataList.ToImmutableArray());");
+
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("}");
+        }
+
+        /// <summary>
+        /// Checks that a candidate method has a Function attribute then proceeds to create a DefaultFunctionMetadata object.
+        /// </summary>
+        private static void AddFunctionMetadataInfo(IndentedTextWriter indentedTextWriter, SyntaxReceiver receiver, Compilation compilation)
+        {
+            var assemblyName = compilation.Assembly.Name;
+            var scriptFile = Path.Combine(assemblyName + ".dll");
+
+            // Loop through the candidate methods (methods with any attribute associated with them)
+            foreach (MethodDeclarationSyntax method in receiver.CandidateMethods)
+            {
+                var model = compilation.GetSemanticModel(method.SyntaxTree);
+                
+                if(!IsMethodAFunction(model, method))
+                {
+                    continue;
+                }
+
+                var functionClass = (ClassDeclarationSyntax)method.Parent!;
+                var functionName = functionClass.Identifier.ValueText;
+                var entryPoint = assemblyName + "." + functionName + "." + method.Identifier.ValueText;
+                var bindingsListName = functionName + "RawBindings";
+
+                // collect Bindings
+                indentedTextWriter.WriteLine($"var {bindingsListName} = new List<string>();");
+                AddBindingInfo(indentedTextWriter, method, model, functionName);
+                indentedTextWriter.WriteLine($"var {functionName} = new DefaultFunctionMetadata");
+                indentedTextWriter.WriteLine("{");
+                indentedTextWriter.Indent++;
+                indentedTextWriter.WriteLine("FunctionId = Guid.NewGuid().ToString(),");
```

Comment:
Is this intentional to have the function ID change on each run of the host? We have an opportunity here to make the ID determined at compile time and to remain the same for each run of that build of the app. Is that valuable? Or is the ID changing on each run of the app intentional/necessary?

Reply:
This is what we do in the host and in other Stein workers (like PyStein) - not sure how necessary this is, but if we find a reason to change the convention we can add it in a new PR/iteration.